sound/soc/sof/topology.c

```diff
@@ -2462,15 +2463,161 @@ static const struct snd_soc_tplg_ops sof_dspless_tplg_ops = {
+		while ((p = strstr(p, sub)) != NULL) {
+			size = (size == 0) ? (p - str) + strlen(p + len) + 1 : size - len;
+			memmove(p, p + len, size - (p - str));
+		}
+	}
+	return str;
+}
+
 int snd_sof_load_topology(struct snd_soc_component *scomp, const char *file)
 {
 	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(scomp);
+	struct snd_sof_pdata *sof_pdata = sdev->pdata;
+	struct topology_file tplg_files[MAX_TPLG_NUM];
+	struct snd_soc_dai_link *dai_link;
 	const struct firmware *fw;
+	bool load_default_tplg = false;
+	unsigned long tplg_mask = 0;
+	char platform[4];
+	char *tplg_name;
+	int tplg_num = 0;
+	int tplg_dev;
 	int ret;
+	int i;
 
 	dev_dbg(scomp->dev, "loading topology:%s\n", file);
 
-	ret = request_firmware(&fw, file, scomp->dev);
+	tplg_name = (char *)file;
+	if (sdev->pdata->ipc_type == SOF_IPC_TYPE_3)
+		goto legacy_tplg;
+
+	sscanf(sof_pdata->tplg_filename, "sof-%3s-*.tplg", platform);
+	if (!is_platform_support_separated_tplg(platform))
+		goto legacy_tplg;
+
+	for_each_card_prelinks(scomp->card, i, dai_link) {
+		char *tplg_device;
+
+		dev_dbg(scomp->dev, "dai_link %s id %d\n", dai_link->name, dai_link->id);
+		if (strstr(dai_link->name, "SimpleJack")) {
+			tplg_dev = TPLG_DEVICE_SDW_JACK;
+			tplg_device = "sdca-jack";
+		} else if (strstr(dai_link->name, "SmartAmp")) {
+			tplg_dev = TPLG_DEVICE_SDW_AMP;
+			tplg_device = devm_kasprintf(sdev->dev, GFP_KERNEL,
+						     "sdca-%damp", dai_link->num_cpus);
+		} else if (strstr(dai_link->name, "SmartMic")) {
+			tplg_dev = TPLG_DEVICE_SDW_MIC;
+			tplg_device = "sdca-mic";
+		} else if (strstr(dai_link->name, "dmic")) {
+			if (strstr(file, "-2ch")) {
+				tplg_device = "dmic-2ch";
+				tplg_name = strremove(tplg_name, "-2ch");
+			} else if (strstr(file, "-4ch")) {
+				tplg_device = "dmic-4ch";
+				tplg_name = strremove(tplg_name, "-4ch");
+			} else {
+				dev_warn(scomp->dev,
+					 "only -2ch and -4ch are supported for dmic\n");
+				continue;
+			}
+			tplg_dev = TPLG_DEVICE_HOST_DMIC;
+		} else if (strstr(dai_link->name, "iDisp")) {
+			tplg_dev = TPLG_DEVICE_HDMI;
+			if (strstr(file, "hda-generic")) {
+				tplg_device = "idisp";
+			} else {
+				tplg_device = "sdca-hdmi";
+			}
+		} else {
+			/* The dai link is not supported by sperated tplg yet */
+			load_default_tplg = true;
+			continue;
+		}
+		if (tplg_mask & BIT(tplg_dev))
+			continue;
+		tplg_mask |= BIT(tplg_dev);
+		tplg_files[tplg_num].be_id = dai_link->id;
+		tplg_files[tplg_num].device = tplg_device;
+		tplg_num++;
+	}
+	dev_dbg(scomp->dev, "tplg_mask %#lx tplg_num %d\n", tplg_mask, tplg_num);
+
+	for (i = 0; i < tplg_num; i++) {
+		tplg_files[i].file = devm_kasprintf(sdev->dev, GFP_KERNEL,
+						    "%s/sof-%s-%s-id%d.tplg",
+						    sof_pdata->tplg_filename_prefix, platform,
+						    tplg_files[i].device,
+						    tplg_files[i].be_id);
+		dev_dbg(scomp->dev, "Requesting %d %s\n", i, tplg_files[i].file);
+		ret = request_firmware(&fw, tplg_files[i].file, scomp->dev);
```

Comment:
the `sof_create_ipc_file_profile()` will check if the firmware and the topology file (the monolithic one!) is in place.
The means that you always need to have the monolithic tplg _and_ the split components installed, otherwise the profile is rejected and if not fallback IPC version is available then we will fail.

I have a feeling that this split tplg handling has to touch the fw-file-profile.c as well.

Reply:
Yeah, I am thinking this open, too. My idea is that we don't touch the monolithic topology. I.e. the monolithic topology will coexist with the split topologies. So that sof_create_ipc_file_profile() will still valid.